src/main.rs

```diff
@@ -342,8 +349,67 @@ impl Display for MenuItem {
     }
 }
 
+fn write_undo_renames(backup_file: PathBuf, replacements: Vec<Rename>) -> anyhow::Result<()> {
+    let undo_replacements = replacements
+        .into_iter()
+        .map(|r| {
+            // make paths absolute to that undo does not depend on CWD
+            let original = if r.original.is_relative() {
+                std::env::current_dir()?.join(r.original)
+            } else {
+                r.original
+            };
+            let new = if r.new.is_relative() {
+                std::env::current_dir()?.join(r.new)
+            } else {
+                r.new
+            };
+
+            Ok(Rename {
+                // swap original and new to get undo replacements
+                original: new,
+                new: original,
+            })
+        })
+        .collect::<anyhow::Result<Vec<_>>>()?;
+
+    let file = fs::File::create(backup_file)?;
+    serde_json::to_writer(file, &undo_replacements)?;
+    Ok(())
+}
+
+fn load_undo_renames(backup_file: PathBuf) -> anyhow::Result<Vec<Rename>> {
+    let file = fs::File::open(&backup_file);
+    let file = match file {
+        Ok(f) => f,
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+            return Err(anyhow!("No undo information found."))
+        }
+        Err(e) => return Err(e.into()),
+    };
+    let undo_replacements: Vec<Rename> = serde_json::from_reader(file)?;
+    for replacement in &undo_replacements {
+        if !replacement.original.exists() {
+            return Err(anyhow!(
+                "Undo not possible. \"{}\" is missing.",
+                replacement.original.display()
+            ));
+        }
+    }
+    fs::remove_file(backup_file)?;
+    Ok(undo_replacements)
+}
+
 fn main() -> anyhow::Result<()> {
     let opts = Opts::parse_from(wild::args());
+    let backup_file = std::env::temp_dir().join("pipe-renamer_undo.json");
+
+    if opts.undo {
+        let replacements = load_undo_renames(backup_file)?;
+        execute_renames(&replacements, opts.rename_command)?;
+        return Ok(());
```

Comment:
optional: Do you want to add some output here for when the undo completes successfully?